README.md
[![Build Status](https://travis-ci.org/stphnlyd/perl5-Math-LOESS.svg?branch=master)](https://travis-ci.org/stphnlyd/perl5-Math-LOESS)
[![AppVeyor Status](https://ci.appveyor.com/api/projects/status/github/stphnlyd/perl5-Math-LOESS?branch=master&svg=true)](https://ci.appveyor.com/project/stphnlyd/perl5-Math-LOESS)

# NAME

Math::LOESS - Perl wrapper of the Locally-Weighted Regression package originally written by Cleveland, et al.

# VERSION

version 0.0001

# SYNOPSIS

```perl
use Math::LOESS;

my $loess = Math::LOESS->new(x => $x, y => $y);

$loess->fit();
my $fitted_values = $loess->outputs->fitted_values;

print $loess->summary();

my $prediction = $loess->predict($new_data, 1);
my $confidence_intervals = $prediction->confidence(0.05);
print $confidence_internals->{fit};
print $confidence_internals->{upper};
print $confidence_internals->{lower};
```

# CONSTRUCTION

```
new((Piddle1D|Piddle2D) :$x, Piddle1D :$y, Piddle1D :$weights=undef,
    Num :$span=0.75, Str :$family='gaussian')
```

Arguments:

- $x

    A `($n, $p)` piddle for x data, where `$p` is number of predictors.
    It's possible to have at most 8 predictors.

- $y

    A `($n, 1)` piddle for y data. 

- $weights

    Optional `($n, 1)` piddle for weights to be given to individual
    observations.
    By default, an unweighted fit is carried out (all the weights are one).

- $span

    The parameter controls the degree of smoothing. Default is 0.75.

    For `span` < 1, the neighbourhood used for the fit includes proportion
    `span` of the points, and these have tricubic weighting (proportional to
    `(1 - (dist/maxdist)^3)^3)`. For `span` > 1, all points are used, with
    the "maximum distance" assumed to be `span^(1/p)` times the actual
    maximum distance for p explanatory variables.

    When provided as a construction parameter, it is like a shortcut for,

    ```
    $loess->model->span($span);
    ```

- $family

    If `"gaussian"` fitting is by least-squares, and if `"symmetric"` a
    re-descending M estimator is used with Tukey's biweight function.

    When provided as a construction parameter, it is like a shortcut for,

    ```
    $loess->model->family($family);
    ```

Bad values in `$x`, `$y`, `$weights` are removed.

# NAME

Math::LOESS - Perl wrapper of the Locally-Weighted Regression package originally written by Cleveland, et al.

# ATTRIBUTES

## model

Get an [Math::LOESS::Model](https://metacpan.org/pod/Math::LOESS::Model) object.

## outputs

Get an [Math::LOESS::Outputs](https://metacpan.org/pod/Math::LOESS::Outputs) object.

## x

Get input x data as a piddle.

## y

Get input y data as a piddle.

## weights

Get input weights data as a piddle.

## activated

Returns a true value if the object's `fit()` method has been called.

# METHODS

## fit

```
fit()
```

## predict

```
predict((Piddle1D|Piddle2D) $newdata, Bool $stderr=false)
```

Returns a [Math::LOESS::Prediction](https://metacpan.org/pod/Math::LOESS::Prediction) object.

Bad values in `$newdata` are removed.

## summary

```
summary()
```

Returns a summary string. For example, 

```
print $loess->summary();
```

# SEE ALSO

[https://en.wikipedia.org/wiki/Local\_regression](https://en.wikipedia.org/wiki/Local_regression)

[PDL](https://metacpan.org/pod/PDL)

# AUTHOR

Stephan Loyd <sloyd@cpan.org>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2019 by Stephan Loyd.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for Math::LOESS

0.0000_04   2019-05-24
  - Change the way of specifying multiple predictors's data.

0.0000_03   2019-05-22
  - Really support multiple predictors now.

0.0000_02   2019-05-21
  - Some fixes and improvements.

0.0000_01   2019-05-20
  - First trail release.



loess/loess.c
#include "S.h"
#include "loess.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>


static char *surf_stat;

int error_status = 0;
char *error_message = NULL;

/* Declarations */

static void
loess_(double *y, double *x_, int *size_info, double *weights, double *span,
       int *degree, int *parametric, int *drop_square, int *normalize,
       char **statistics, char **surface, double *cell, char **trace_hat_in,
       int *iterations, double *fitted_values, double *fitted_residuals,
       double *enp, double *residual_scale, double *one_delta, double *two_delta,
       double *pseudovalues, double *trace_hat_out, double *diagonal,
       double *robust, double *divisor, int *parameter, int *a, double *xi,
       double *vert, double *vval);

void F77_SUB(lowesw)(double*, int*, double*, double*);
void F77_SUB(lowesp)(int*, double*, double*, double*,
                     double*, double*, double*);

static void
condition(char **surface, char *new_stat, char **trace_hat_in)
{
    if(!strcmp(*surface, "interpolate")) {
        if(!strcmp(new_stat, "none"))
            surf_stat = "interpolate/none";
        else if(!strcmp(new_stat, "exact"))
            surf_stat = "interpolate/exact";
        else if(!strcmp(new_stat, "approximate"))
        {
            if(!strcmp(*trace_hat_in, "approximate"))
                surf_stat = "interpolate/2.approx";
            else if(!strcmp(*trace_hat_in, "exact"))
                surf_stat = "interpolate/1.approx";
        }
    }
    else if(!strcmp(*surface, "direct")) {
        if(!strcmp(new_stat, "none"))
            surf_stat = "direct/none";
        else if(!strcmp(new_stat, "exact"))
            surf_stat = "direct/exact";
        else if(!strcmp(new_stat, "approximate"))
            surf_stat = "direct/approximate";
    }
}

static int
comp(const void *d1, const void *d2)
{
   double *_d1 = (double *)d1;
   double *_d2 = (double *)d2;

    if(*_d1 < *_d2)
        return(-1);
    else if(*_d1 == *_d2)
        return(0);
    else
        return(1);
}


void
loess_model_setup(loess_model *model) {
    int i;

    model->span = 0.75;
    model->degree = 2;
    model->normalize = TRUE;

    for(i = 0; i < 8; i++) {
        model->parametric[i] = FALSE;
        model->drop_square[i] = FALSE;
    }

    model->family = "gaussian";
}

void
loess_inputs_setup(double *x, double *y, double *w, long n,
      long p, loess_inputs *inputs) {
    int i;

    inputs->y = MALLOC(n * sizeof(double));
    inputs->x = MALLOC(n * p * sizeof(double));
    inputs->weights = MALLOC(n * sizeof(double));

    for(i = 0; i < (n * p); i++) {
        inputs->x[i] = x[i];
    }

    for(i = 0; i < n; i++) {
        inputs->y[i] = y[i];
    }

    if (w == NULL) {
        for(i = 0; i < n; i++) {
            inputs->weights[i] = 1.0;
        }
    } else {
        for(i = 0; i < n; i++) {
            inputs->weights[i] = w[i];
        }
    }

    inputs->n = n;
    inputs->p = p;
}

void loess_outputs_setup(long n, long p, loess_outputs *outputs) {
   outputs->fitted_values = MALLOC(n * sizeof(double));
   outputs->fitted_residuals = MALLOC(n * sizeof(double));
   outputs->diagonal = MALLOC(n * sizeof(double));
   outputs->robust = MALLOC(n * sizeof(double));
   outputs->divisor = MALLOC(p * sizeof(double));
   outputs->pseudovalues = MALLOC(n * sizeof(double));
}

void
loess_kd_tree_setup(long n, long p, loess_kd_tree *kd_tree) {
   int max_kd;

   max_kd = n > 200 ? n : 200;

   kd_tree->parameter = MALLOC(7 * sizeof(int));
   kd_tree->a = MALLOC(max_kd * sizeof(int));
   kd_tree->xi = MALLOC(max_kd * sizeof(double));
   kd_tree->vert = MALLOC(p * 2 * sizeof(double));
   kd_tree->vval = MALLOC((p + 1) * max_kd * sizeof(double));
}

void
loess_control_setup(loess_control *control) {
   control->surface = "interpolate";
   control->statistics = "approximate";
   control->cell = 0.2;
   control->trace_hat = "wait.to.decide";
   control->iterations = 4;
}

void
loess_setup(double *x, double *y, double *w, long n, long p, loess *lo)
{
   loess_inputs_setup(x, y, w, n, p, lo->inputs);
   loess_model_setup(lo->model);
   loess_control_setup(lo->control);
   loess_outputs_setup(n, p, lo->outputs);
   loess_kd_tree_setup(n, p, lo->kd_tree);
}

static void
loess_(double *y, double *x_, int *size_info, double *weights, double *span,
       int *degree, int *parametric, int *drop_square, int *normalize,
       char **statistics, char **surface, double *cell, char **trace_hat_in,
       int *iterations, double *fitted_values, double *fitted_residuals,
       double *enp, double *residual_scale, double *one_delta, double *two_delta,
       double *pseudovalues, double *trace_hat_out, double *diagonal,
       double *robust, double *divisor, int *parameter, int *a, double *xi,
       double *vert, double *vval)
{
    double  *x, *x_tmp, new_cell, trL, delta1, delta2, sum_squares = 0,
            pseudo_resid, *temp, *xi_tmp, *vert_tmp, *vval_tmp,
            *diag_tmp, trL_tmp = 0, d1_tmp = 0, d2_tmp = 0, sum, mean;
    int    i, j, k, p, N, D, sum_drop_sqr = 0, sum_parametric = 0,
            setLf, nonparametric = 0, *order_parametric,
            *order_drop_sqr, zero = 0, max_kd;
    int *param_tmp, *a_tmp;
    int     cut;
    char    *new_stat;

    D = size_info[0];
    N = size_info[1];
    max_kd = (N > 200 ? N : 200);
    *one_delta = *two_delta = *trace_hat_out = 0;

    x = MALLOC(D * N * sizeof(double));
    x_tmp = MALLOC(D * N * sizeof(double));
    temp = MALLOC(N * sizeof(double));
    a_tmp = MALLOC(max_kd * sizeof(int));
    xi_tmp = MALLOC(max_kd * sizeof(double));
    vert_tmp = MALLOC(D * 2 * sizeof(double));
    vval_tmp = MALLOC((D + 1) * max_kd * sizeof(double));
    diag_tmp = MALLOC(N * sizeof(double));
    param_tmp = MALLOC(N * sizeof(int));
    order_parametric = MALLOC(D * sizeof(int));
    order_drop_sqr = MALLOC(D * sizeof(int));

    new_cell = (*span) * (*cell);
    for(i = 0; i < N; i++)
        robust[i] = 1;
        for(i = 0; i < (N * D); i++)
            x_tmp[i] = x_[i];
    if((*normalize) && (D > 1)) {
        cut = ceil(0.100000000000000000001 * N);
        for(i = 0; i < D; i++) {
            k = i * N;
            for(j = 0; j < N; j++)
                temp[j] = x_[k + j];
            qsort(temp, N, sizeof(double), comp);
            sum = 0;
            for(j = cut; j <= (N - cut - 1); j++)
                sum = sum + temp[j];
            mean = sum / (N - 2 * cut);
            sum = 0;
            for(j = cut; j <= (N - cut - 1); j++) {
                temp[j] = temp[j] - mean;
                sum = sum + temp[j] * temp[j];
            }
            divisor[i] = sqrt(sum / (N - 2 * cut - 1));
            for(j = 0; j < N; j++) {
                p = k + j;
                x_tmp[p] = x_[p] / divisor[i];
            }
        }
    }
    else
        for(i = 0; i < D; i++) divisor[i] = 1;

    j = D - 1;
    for(i = 0; i < D; i++) {
        sum_drop_sqr = sum_drop_sqr + drop_square[i];
        sum_parametric = sum_parametric + parametric[i];
        if(parametric[i])
            order_parametric[j--] = i;
        else
            order_parametric[nonparametric++] = i;
    }
    //Reorder the predictors w/ the non-parametric first
    for(i = 0; i < D; i++) {
        order_drop_sqr[i] = 2 - drop_square[order_parametric[i]];
        k = i * N;
        p = order_parametric[i] * N;
        for(j = 0; j < N; j++)
            x[k + j] = x_tmp[p + j];
    }

    // Misc. checks .............................
    if((*degree) == 1 && sum_drop_sqr) {
    	error_status = 1;
    	error_message = "Specified the square of a factor predictor to be "\
                        "dropped when degree = 1";
        return;
    }

    if(D == 1 && sum_drop_sqr) {
    	error_status = 1;
        error_message = "Specified the square of a predictor to be dropped "\
                         "with only one numeric predictor";
        return;
    }

    if(sum_parametric == D) {
    	error_status = 1;
        error_message = "Specified parametric for all predictors";
        return;
    }

    // Start the iterations .....................
    for(j = 0; j <= (*iterations); j++) {
        new_stat = j ? "none" : *statistics;
        for(i = 0; i < N; i++)
            robust[i] = weights[i] * robust[i];
        condition(surface, new_stat, trace_hat_in);
        setLf = !strcmp(surf_stat, "interpolate/exact");
        loess_raw(y, x, weights, robust, &D, &N, span, degree,
                  &nonparametric, order_drop_sqr, &sum_drop_sqr,
                  &new_cell, &surf_stat, fitted_values, parameter, a,
                  xi, vert, vval, diagonal, &trL, &delta1, &delta2,
                  &setLf);
        if(j == 0) {
            *trace_hat_out = trL;
            *one_delta = delta1;
            *two_delta = delta2;
        }
        for(i = 0; i < N; i++){
            fitted_residuals[i] = y[i] - fitted_values[i];
        };
        if(j < (*iterations))
            F77_SUB(lowesw)(fitted_residuals, &N, robust, temp);
    }

    if((*iterations) > 0) {
        F77_SUB(lowesp)(&N, y, fitted_values, weights, robust, temp,
						pseudovalues);
        loess_raw(pseudovalues, x, weights, weights, &D, &N, span,
                  degree, &nonparametric, order_drop_sqr, &sum_drop_sqr,
                  &new_cell, &surf_stat, temp, param_tmp, a_tmp, xi_tmp,
                  vert_tmp, vval_tmp, diag_tmp, &trL_tmp, &d1_tmp, &d2_tmp,
                  &zero);
        for(i = 0; i < N; i++) {
            pseudo_resid = pseudovalues[i] - temp[i];
            sum_squares = sum_squares + weights[i] * pseudo_resid * pseudo_resid;
        }
    } else {
        for(i = 0; i < N; i++)
            sum_squares = sum_squares + weights[i] *
                    fitted_residuals[i] * fitted_residuals[i];
    }

    *enp = (*one_delta) + 2 * (*trace_hat_out) - N;
    *residual_scale = sqrt(sum_squares / (*one_delta));

    //Clean the mess and leave ..................
    free(x);
    free(x_tmp);
    free(temp);
    free(xi_tmp);
    free(vert_tmp);
    free(vval_tmp);
    free(diag_tmp);
    free(a_tmp);
    free(param_tmp);
    free(order_parametric);
    free(order_drop_sqr);
}

void
loess_fit(loess *lo)
{
    int    size_info[2], iterations;

    size_info[0] = lo->inputs->p;
    size_info[1] = lo->inputs->n;

    //Reset the default error status...
    error_status = 0;
    lo->status.err_status = 0;
    lo->status.err_msg = NULL;

    iterations = (!strcmp(lo->model->family, "gaussian")) ? 0 :
    lo->control->iterations;
    if(!strcmp(lo->control->trace_hat, "wait.to.decide")) {
        if(!strcmp(lo->control->surface, "interpolate"))
            lo->control->trace_hat = (lo->inputs->n < 500) ? "exact" : "approximate";
        else
            lo->control->trace_hat = "exact";
    }
    loess_(lo->inputs->y, lo->inputs->x, size_info, lo->inputs->weights,
           &lo->model->span,
           &lo->model->degree,
           lo->model->parametric,
           lo->model->drop_square,
           &lo->model->normalize,
           &lo->control->statistics,
           &lo->control->surface,
           &lo->control->cell,
           &lo->control->trace_hat,
           &iterations,
           lo->outputs->fitted_values,
           lo->outputs->fitted_residuals,
           &lo->outputs->enp,
           &lo->outputs->residual_scale,
           &lo->outputs->one_delta,
           &lo->outputs->two_delta,
           lo->outputs->pseudovalues,
           &lo->outputs->trace_hat,
           lo->outputs->diagonal,
           lo->outputs->robust,
           lo->outputs->divisor,
           lo->kd_tree->parameter,
           lo->kd_tree->a,
           lo->kd_tree->xi,
           lo->kd_tree->vert,
           lo->kd_tree->vval);

    if(error_status){
        lo->status.err_status = error_status;
        lo->status.err_msg = error_message;
    }
}


void loess_inputs_free(loess_inputs *inputs) {
   free(inputs->x);
   free(inputs->y);
   free(inputs->weights);
}

void loess_outputs_free(loess_outputs *outputs) {
   free(outputs->fitted_values);
   free(outputs->fitted_residuals);
   free(outputs->diagonal);
   free(outputs->robust);
   free(outputs->divisor);
   free(outputs->pseudovalues);
}

void loess_kd_tree_free(loess_kd_tree *kd_tree) {
   free(kd_tree->parameter);
   free(kd_tree->a);
   free(kd_tree->xi);
   free(kd_tree->vert);
   free(kd_tree->vval);
}

void
loess_free_mem(loess *lo)
{
   loess_inputs_free(lo->inputs);
   loess_outputs_free(lo->outputs);
   loess_kd_tree_free(lo->kd_tree);
}

/*
void
loess_summary(loess *lo)
{
    printf("Number of Observations         : %ld\n", lo->inputs->n);
    printf("Equivalent Number of Parameters: %.1f\n", lo->outputs->enp);
    if(!strcmp(lo->model->family, "gaussian"))
        printf("Residual Standard Error        : ");
    else
        printf("Residual Scale Estimate: ");
    printf("%.4f\n", lo->outputs->residual_scale);
}
*/


lib/Math/LOESS.pm
package Math::LOESS;

# ABSTRACT: Perl wrapper of the Locally-Weighted Regression package originally written by Cleveland, et al.

use 5.010;
use strict;
use warnings;

# VERSION

use List::Util qw(reduce);
use PDL::Core qw(ones);
use PDL::Lite;
use Math::LOESS::_swig;
use Type::Params qw(compile compile_named);
use Types::Standard qw(Any Num Object Optional Str);
use Types::PDL qw(Piddle1D Piddle2D);

use Math::LOESS::Model;
use Math::LOESS::Outputs;
use Math::LOESS::Prediction;

sub new {
    my $class = shift;
    state $check = compile_named(
        x       => (Piddle1D | Piddle2D),
        y       => Piddle1D,
        weights => Optional [Piddle1D],
        span    => Optional [ Num->where( sub { $_ > 0 and $_ < 1 } ) ],
        family  => Optional [ Str ],
    );

    my $arg = $check->(@_);

    my ( $x, $y, $span, $family ) = map { $arg->{$_} } qw(x y span family);

    my $n = $y->dim(0);
    if ($x->dim(0) != $n) {
        die "x and y data must have same length at dim 0";
    }
    my $p = $x->dim(1);
    if ($p > 8) {
        die "Cannot have more than 8 ($p seen) predictors.";
    }

    my $w = $arg->{weights};
    if ( defined $w ) {
        unless ( $w->dim(0) == $n ) {
            die "weights if given must have same length as y";
        }
        ($x, $y, $w) = $class->_process_bad_values($x, $y, $w);
    }
    else {
        ($x, $y) = $class->_process_bad_values($x, $y);
    }
    $n = $y->dim(0);

    my $x1 = Math::LOESS::_swig::pdl_to_darray($x);
    my $y1 = Math::LOESS::_swig::pdl_to_darray($y);
    my $w1 = defined $w ? Math::LOESS::_swig::pdl_to_darray($w) : undef;

    my $loess = Math::LOESS::_swig::loess->new();
    for my $param (qw(inputs model control kd_tree outputs)) {
        my $class = "Math::LOESS::_swig::loess_${param}";
        $loess->{$param} = $class->new();
    }
    
    my $self = bless(
        {
            %$arg,
            _obj      => $loess,
            activated => 0
        },
        $class
    );

    Math::LOESS::_swig::loess_setup( $x1, $y1, $w1, $n, $p, $self->_obj );

    # free memory
    map { Math::LOESS::_swig::delete_doubleArray($_) }
      ( $x1, $y1, defined $w1 ? $w1 : () );

    $self->{model} = Math::LOESS::Model->new( _obj => $self->_obj->{model} );
    if ( defined $span ) {
        $self->model->span($span);
    }
    if ( defined $family ) {
        $self->model->family($family);
    }

    $self->{outputs} = Math::LOESS::Outputs->new(
        _obj   => $self->_obj->{outputs},
        _loess => $self,
    );

    return $self;
}

sub DESTROY {
    my ($self) = @_;
    Math::LOESS::_swig::loess_free_mem( $self->_obj );
}

for my $attr (qw(_obj activated model outputs)) {
    no strict 'refs';
    *{$attr} = sub { $_[0]->{$attr} };
}

sub _inputs   { $_[0]->_obj->{inputs} }
sub n { $_[0]->_inputs->{n} }
sub p { $_[0]->_inputs->{p} }

sub _size {
    my ( $self, $attr ) = @_;
    return $attr eq 'x'
      ? ($self->n, $self->p)
      : ($self->n, 1);
}

for my $attr (qw(x y weights)) {
    no strict 'refs';
    *{$attr} = sub {
        my ($self) = @_;

        my $d = $self->_obj->{inputs}->{$attr};
        return Math::LOESS::_swig::darray_to_pdl( $d, $self->_size($attr) );
    };
}

sub _check_error {
    my ($self) = @_;

    my $status = $self->_obj->{status};
    if ( $status->{err_status} ) {
        die $status->{err_msg};
    }
}

sub fit {
    my ($self) = @_;

    Math::LOESS::_swig::loess_fit( $self->_obj );
    $self->_check_error;
    $self->{activated} = 1;
    return;
}

sub predict {
    state $check = compile( Any, (Piddle1D | Piddle2D), Optional [Any] );
    my ($self, $newdata, $stderr) = $check->(@_);

    unless ( $newdata->dim(1) == $self->p ) {
        die "newdata's dim 1 must be same as that of x";
    }
    ($newdata) = $self->_process_bad_values($newdata);

    unless ($self->activated) {
        $self->fit();
    }

    my $pred = Math::LOESS::_swig::prediction->new();
    $pred->{m} = $newdata->dim(0);
    $pred->{se} = $stderr ? 1 : 0; 

    my $d = Math::LOESS::_swig::pdl_to_darray($newdata);
    Math::LOESS::_swig::predict( $d, $self->_obj, $pred );
    Math::LOESS::_swig::delete_doubleArray($d);
    $self->_check_error;

    return Math::LOESS::Prediction->new( _obj => $pred );
}

sub summary {
    my ($self) = @_;

    my $fit_flag = $self->activated;
    my $s = sprintf(
        <<"EOT",
Number of Observations         : %d
Fit flag                       : %d
Equivalent Number of Parameters: %.1f
EOT
        $self->n, $fit_flag, $self->outputs->enp
    );

    if ($fit_flag) {
        my $rse_fmt =
          $self->model->family eq 'gaussian'
          ? "Residual Standard Error        : %.4f"
          : "Residual Scale Estimate        : %.4f";
        $s .= sprintf( $rse_fmt, $self->outputs->residual_scale ) . "\n";
    }
    return $s;
}

sub _process_bad_values {
    my ( $class, $p1, @rest ) = @_;    # $p1 can be multi dimensional

    if ( $p1->badflag or ( grep { $_->badflag } @rest ) ) {
        my $p1_isbad = $p1->isbad;
        my @p1_isbad =
          map { $p1_isbad->slice(",($_)") } ( 0 .. $p1->dim(1) - 1 );
        my $isbad =
          reduce { ( $a | $b ) } ( @p1_isbad, map { $_->isbad } @rest );
        return ( map { $_->where( !$isbad ) } ( $p1, @rest ) );
    }
    return ( $p1, @rest );
}

1;

__END__

=head1 NAME

Math::LOESS - Perl wrapper of the Locally-Weighted Regression package originally written by Cleveland, et al.

=head1 SYNOPSIS

    use Math::LOESS;

    my $loess = Math::LOESS->new(x => $x, y => $y);

    $loess->fit();
    my $fitted_values = $loess->outputs->fitted_values;

    print $loess->summary();

    my $prediction = $loess->predict($new_data, 1);
    my $confidence_intervals = $prediction->confidence(0.05);
    print $confidence_internals->{fit};
    print $confidence_internals->{upper};
    print $confidence_internals->{lower};

=head1 CONSTRUCTION

    new((Piddle1D|Piddle2D) :$x, Piddle1D :$y, Piddle1D :$weights=undef,
        Num :$span=0.75, Str :$family='gaussian')

Arguments:

=over 4

=item * $x

A C<($n, $p)> piddle for x data, where C<$p> is number of predictors.
It's possible to have at most 8 predictors.

=item * $y

A C<($n, 1)> piddle for y data. 

=item * $weights

Optional C<($n, 1)> piddle for weights to be given to individual
observations.
By default, an unweighted fit is carried out (all the weights are one).

=item * $span

=include span@Math::LOESS::Model

When provided as a construction parameter, it is like a shortcut for,

    $loess->model->span($span);

=item * $family

=include family@Math::LOESS::Model

When provided as a construction parameter, it is like a shortcut for,

    $loess->model->family($family);

=back

Bad values in C<$x>, C<$y>, C<$weights> are removed.

=head1 ATTRIBUTES

=head2 model

Get an L<Math::LOESS::Model> object.

=head2 outputs

Get an L<Math::LOESS::Outputs> object.

=head2 x

Get input x data as a piddle.

=head2 y

Get input y data as a piddle.

=head2 weights

Get input weights data as a piddle.

=head2 activated

Returns a true value if the object's C<fit()> method has been called.

=head1 METHODS

=head2 fit

    fit()

=head2 predict

    predict((Piddle1D|Piddle2D) $newdata, Bool $stderr=false)

Returns a L<Math::LOESS::Prediction> object.

Bad values in C<$newdata> are removed.

=head2 summary

    summary()

Returns a summary string. For example, 

    print $loess->summary();

=head1 SEE ALSO

L<https://en.wikipedia.org/wiki/Local_regression>

L<PDL>



lib/Math/LOESS/Outputs.pm
package Math::LOESS::Outputs;

# ABSTRACT: Math::LOESS basic fit outputs

use 5.010;
use strict;
use warnings;

# VERSION

use Math::LOESS::_swig;
use Scalar::Util qw(weaken);
use Type::Params qw(compile_named);
use Types::Standard qw(Int Object Str);

sub new {
    my $class = shift;
    state $check = compile_named(
        _obj   => Object,
        _loess => Object,
    );

    my $arg = $check->(@_);
    my $self = bless( $arg, $class );
    weaken( $self->{_loess} );
    return $self;
}

sub _loess { $_[0]->{_loess} }

sub _obj {
    my ($self) = @_;
    unless ($self->_loess) {
        die 'Math::LOESS::Outputs object has been invalidated '
          . 'because its corresponding loess object has been destroyed. '
          . 'Make sure the loess object be living when calling '
          . 'Math::LOESS::Outputs object methods.'
    }
    return $self->{_obj};
}

sub _family { $_[0]->_loess->model->family }
sub _n { $_[0]->_loess->n }
sub _p { $_[0]->_loess->p }

for my $attr (
    qw(
    fitted_values fitted_residuals diagnal robust
    divisor
    )
  )
{
    no strict 'refs';
    *{$attr} = sub {
        my ($self) = @_;
        return Math::LOESS::_swig::darray_to_pdl( $self->_obj->{$attr},
            $self->_n );
    };
}

sub pseudovalues {
    my ($self) = @_;
    if ( $self->family ne 'symmetric' ) {
        die "pseudovalues are available only when robust fitting. "
          . "Use family='symmetric' for robust fitting";
    }
    return Math::LOESS::_swig::darray_to_pdl( $self->_obj->{pseudovalues},
        $self->_n );
}

for my $attr (
    qw(
    enp residual_scale one_delta two_delta trace_hat
    )
  )
{
    no strict 'refs';
    *{$attr} = sub {
        my ($self) = @_;
        return $self->_obj->{$attr};
    };
}

1;

__END__

=head1 DESCRIPTION

You normally don't need to construct object of this class yourself.
Instead you get the object from an L<Math::LOESS> object after its C<fit()>
method is called. 

=head1 ATTRIBUTES

=head2 fitted_values

Fitted values.

=head2 fitted_residuals

Fitted residuals.

=head2 pseudovalues

Adjusted values of the response when robust estimation is used.

=head2 diagnal

Diagonal of the operator hat matrix.

=head2 robust

Robustness weights for robust fitting.

=head2 divisor

Normalization divisors for numeric predictors.

=head2 enp

Equivalent number of parameters.

=head2 residual_scale

Estimate of the scale of residuals.

=head2 one_delta

Statistical parameter used in the computation of standard errors.

=head2 two_delta

Statistical parameter used in the computation of standard errors.

=head2 trace_hat

Trace of the operator hat matrix.

=head1 SEE ALSO

L<Math::LOESS>



lib/Math/LOESS/Prediction.pm
package Math::LOESS::Prediction;

# ABSTRACT: Math::LOESS prediction and confidence intervals

use 5.010;
use strict;
use warnings;

# VERSION

use Math::LOESS::_swig;
use Type::Params qw(compile_named);
use Types::Standard qw(Object);

sub new {
    my $class = shift;
    state $check = compile_named( _obj => Object );

    my $arg = $check->(@_);
    return bless( $arg, $class );
}

sub DESTROY {
    my ($self) = @_;
    Math::LOESS::_swig::pred_free_mem( $self->_obj );
}

sub _obj { $_[0]->{_obj} };

sub values {
    my ($self) = @_;
    return Math::LOESS::_swig::darray_to_pdl( $self->_obj->{fit},
        $self->_obj->{m} );
}

sub stderr {
    my ($self) = @_;
    unless ( $self->_obj->{se} ) {
        die "Standard error was not computed. "
          . "Use stderr => 1 when predicting";
    }
    return Math::LOESS::_swig::darray_to_pdl( $self->_obj->{se_fit},
        $self->_obj->{m} );
}

for my $attr (qw(residual_scale df)) {
    no strict 'refs';
    *{$attr} = sub {
        my ($self) = @_;
        return $self->_obj->{$attr};
    };
}

sub confidence {
    my ($self, $alpha) = @_;
    $alpha //= 0.01;

    unless ($alpha > 0 and $alpha < 1) {
        die "The alpha value should be between 0 and 1";
    }
    unless ($self->_obj->{se}) {
        die "Cannot compute confidence intervals without standard errors";
    }

    my $ci = Math::LOESS::_swig::confidence_intervals->new;
    Math::LOESS::_swig::pointwise($self->_obj, 1 - $alpha, $ci);   
    my $m = $self->_obj->{m};
    my $rslt =
      { map { $_ => Math::LOESS::_swig::darray_to_pdl( $ci->{$_}, $m ) }
          qw(fit upper lower) };
    Math::LOESS::_swig::pw_free_mem($ci);
    return $rslt;
}

1;

__END__

=head1 DESCRIPTION

You normally don't need to construct object of this class yourself.
Instead you get the object from an L<Math::LOESS> object after its C<fit()>
method is called. 

=head1 ATTRIBUTES

=head2 values

loess values evaluated at newdata.

=head2 stderr

Estimates of the standard error on the estimated values.

=head2 residual_scale

Estimate of the scale of the residuals.

=head2 df

Degrees of freedom of the loess fit.

It is used with the t-distribution to compute pointwise confidence
intervals for the evaluated surface. It is obtained using the formula
C<(one_delta ** 2) / two_delta>

=head1 METHODS

=head2 confidence

    confidence($alpha=0.01)

Returns the confidence intervals for predicted values, as a hashref of
C<{ fit =E<gt> $fit, upper =E<gt> $upper, lower =E<gt> $lower }>,
where C<$fit>, C<$upper>, C<$lower> are piddles.

=head1 SEE ALSO

L<Math::LOESS>



lib/Math/LOESS/Model.pm
package Math::LOESS::Model;

# ABSTRACT: Math::LOESS model configurations

use 5.010;
use strict;
use warnings;

our $VERSION = '0.0000_02'; # TRIAL VERSION

use Math::LOESS::_swig;
use Type::Params qw(compile_named);
use Types::Standard qw(ArrayRef Enum Object);

sub new {
    my $class = shift;
    state $check = compile_named( _obj => Object );

    my $arg = $check->(@_);
    return bless( $arg, $class );
}

sub _obj { $_[0]->{_obj} };

for my $attr (qw(span degree normalize)) {
    no strict 'refs';
    *{$attr} = sub {
        my ($self, $value) = @_;

        if (defined $value) {
            $self->_obj->{$attr} = $value;
        }
        return $self->_obj->{$attr};
    };
}

sub family {
    my ($self, $value) = @_;

    if (defined $value) {
        state $check = Type::Params::compile(Enum[qw(gaussian symmetric)]);
        ($value) = $check->($value);
        $self->_obj->{family} = $value;
    }
    return $self->_obj->{family};
}

for my $attr (qw(parametric drop_square)) {
    no strict 'refs';
    *{$attr} = sub {
        my ($self, $value) = @_;

        if (defined $value) {
            $self->_obj->{$attr} = $value;
        }
        return $self->_obj->{$attr};
    };
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Math::LOESS::Model - Math::LOESS model configurations

=head1 VERSION

version 0.0000_02

=head1 DESCRIPTION

You normally don't need to construct object of this class yourself.
Instead you get the object from an L<Math::LOESS> object.

=head1 ATTRIBUTES

=head2 span

=tmpl span

The parameter controls the degree of smoothing. Default is 0.75.

For C<span> < 1, the neighbourhood used for the fit includes proportion
C<span> of the points, and these have tricubic weighting (proportional to
C<(1 - (dist/maxdist)^3)^3)>. For C<span> > 1, all points are used, with
the "maximum distance" assumed to be C<span^(1/p)> times the actual
maximum distance for p explanatory variables.

=tmpl

=head2 degree

The degree of the polynomials to be used, normally 1 or 2. Default is 2.

=head2 parametric

Should any terms be fitted globally rather than locally? Default is false.
Terms can be specified by name, number or as a logical vector of the same
length as the number of predictors.

=head2 drop_square

For fits with more than one predictor and degree = 2, should the quadratic
term be dropped for particular predictors? Default is false.
Terms are specified in the same way as for parametric.

=head2 normalize

Should the predictors be normalized to a common scale if there is more
than one? The normalization used is to set the 10% trimmed standard
deviation to one. Set to false for spatial coordinate predictors and
others known to be on a common scale.

=head2 family

=tmpl family

If C<"gaussian"> fitting is by least-squares, and if C<"symmetric"> a
re-descending M estimator is used with Tukey's biweight function.

=tmpl

=head1 SEE ALSO

L<Math::LOESS>

=head1 AUTHOR

Stephan Loyd <sloyd@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2019 by Stephan Loyd.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
